core/src/main/java/io/grpc/internal/BinaryLog.java

```diff
@@ -0,0 +1,208 @@
+      for (String configuration : configurations) {
+        Matcher configMatcher = configRe.matcher(configuration);
+        if (!configMatcher.matches()) {
+          throw new IllegalArgumentException("Bad input: " + configuration);
+        }
+        String methodOrSvc = configMatcher.group(1);
+        String binlogOptionStr = configMatcher.group(2);
+        BinaryLog binLog = createBinaryLog(binlogOptionStr);
+        if (methodOrSvc.equals("*")) {
+          if (globalLog != null) {
+            throw new IllegalArgumentException("Duplicate log config for: *");
+          }
+          globalLog = binLog;
+          logger.info("Global binlog: " + globalLog);
+        } else if (isServiceGlob(methodOrSvc)) {
+          String service = MethodDescriptor.extractFullServiceName(methodOrSvc);
+          if (perServiceLogs.containsKey(service)) {
+            throw new IllegalArgumentException("Duplicate log config for service: " + methodOrSvc);
+          }
+          perServiceLogs.put(service, binLog);
+          logger.info(String.format("Service binlog: service=%s log=%s", service, binLog));
+        } else {
+          // assume fully qualified method name
+          if (perMethodLogs.containsKey(methodOrSvc)) {
+            throw new IllegalArgumentException("Duplicate log config for method: " + methodOrSvc);
+          }
+          perMethodLogs.put(methodOrSvc, binLog);
+          logger.info(String.format("Method binlog: method=%s log=%s", methodOrSvc, binLog));
+        }
+      }
+      enabled = true;
+      this.globalLog = globalLog == null ? NOOP_LOG : globalLog;
+    }
+
+    /**
+     * Accepts a full method name and returns the log that should be used.
+     */
+    public BinaryLog getLog(String fullMethodName) {
+      if (!enabled) {
+        return NOOP_LOG;
+      }
+      BinaryLog methodLog = perMethodLogs.get(fullMethodName);
+      if (methodLog != null) {
+        return methodLog;
+      }
+      BinaryLog serviceLog = perServiceLogs.get(
+          MethodDescriptor.extractFullServiceName(fullMethodName));
+      if (serviceLog != null) {
+        return serviceLog;
+      }
+      return globalLog;
+    }
+
+    /**
+     * Returns a binlog with the correct header and message limits. The input should be a string
+     * that is in one of these forms:
+     *
+     * <p>{@code {h(:\d+)?}, {m(:\d+)?}, {h(:\d+)?,m(:\d+)?}}
+     *
+     * <p>If the {@code logConfig} is null, the returned binlog will have a limit of
+     * Integer.MAX_VALUE.
+     */
+    @VisibleForTesting
+    static BinaryLog createBinaryLog(@Nullable String logConfig) {
+      if (logConfig == null) {
+        return new BinaryLog(Integer.MAX_VALUE, Integer.MAX_VALUE);
+      }
+      Matcher headerMatcher;
+      Matcher msgMatcher;
+      Matcher bothMatcher;
+      final int maxHeaderBytes;
+      final int maxMsgBytes;
+      if ((headerMatcher = headerRe.matcher(logConfig)).matches()) {
+        maxMsgBytes = 0;
```

Comment:
Would you mind assigning these in the same order each time? So this would be the last line of this block.

Reply:
Done